calico_node/calico-bgp-daemon/main.go

```diff
@@ -90,83 +125,183 @@ func getGlobalConfig(api etcd.KeysAPI) (*bgpconfig.Global, error) {
+			switch key {
+			case "ip_addr_v4":
+				v4 = v.Value
+				if v4 == os.Getenv(IP) {
+					v4 = ""
 				}
-			}
-			if v4 != "" {
-				ns = append(ns, &bgpconfig.Neighbor{
-					Config: bgpconfig.NeighborConfig{
-						NeighborAddress: v4,
-						PeerAs:          peerASN,
-					},
-				})
-			}
-			if v6 != "" {
-				ns = append(ns, &bgpconfig.Neighbor{
-					Config: bgpconfig.NeighborConfig{
-						NeighborAddress: v6,
-						PeerAs:          peerASN,
-					},
-				})
+			case "ip_addr_v6":
+				v6 = v.Value
+				if v6 == os.Getenv(IP6) {
+					v6 = ""
+				}
+			case "as_num":
+				asn, err := strconv.ParseUint(v.Value, 10, 32)
+				if err != nil {
+					return nil, err
+				}
+				peerASN = uint32(asn)
+			default:
+				log.Printf("unhandled key: %s", v.Key)
 			}
 		}
-		return ns, nil
+		if v4 != "" {
+			id := strings.Replace(v4, ".", "_", -1)
+			ns = append(ns, &bgpconfig.Neighbor{
+				Config: bgpconfig.NeighborConfig{
+					NeighborAddress: v4,
+					PeerAs:          peerASN,
+					Description:     fmt.Sprintf("Mesh_%s", id),
+				},
+			})
+		}
+		if v6 != "" {
+			id := strings.Replace(v4, ":", "_", -1)
+			ns = append(ns, &bgpconfig.Neighbor{
+				Config: bgpconfig.NeighborConfig{
+					NeighborAddress: v6,
+					PeerAs:          peerASN,
+					Description:     fmt.Sprintf("Mesh_%s", id),
+				},
+			})
+		}
 	}
-	return nil, nil
+	return ns, nil
+
 }
 
-func makePath(key string, isWithdrawal bool) (*bgptable.Path, error) {
+func getNeighborConfigFromPeer(node *etcd.Node, neighborType string) (*bgpconfig.Neighbor, error) {
+	m := &struct {
+		IP  string `json:"ip"`
+		ASN string `json:"as_num"`
+	}{}
+	if err := json.Unmarshal([]byte(node.Value), m); err != nil {
+		return nil, err
+	}
+	asn, err := strconv.ParseUint(m.ASN, 10, 32)
+	if err != nil {
+		return nil, err
+	}
+	return &bgpconfig.Neighbor{
+		Config: bgpconfig.NeighborConfig{
+			NeighborAddress: m.IP,
+			PeerAs:          uint32(asn),
+			Description:     fmt.Sprintf("%s_%s", strings.Title(neighborType), underscore(m.IP)),
+		},
+	}, nil
+}
 
-	attrs := []bgp.PathAttributeInterface{
-		bgp.NewPathAttributeOrigin(0),
-		bgp.NewPathAttributeNextHop(os.Getenv(IP)),
+func getNonMeshNeighborConfigs(api etcd.KeysAPI, neighborType, version string) ([]*bgpconfig.Neighbor, error) {
+	var key string
+	switch neighborType {
+	case "global":
+		key = fmt.Sprintf("%s/global/peer_%s", CALICO_BGP, version)
+	case "node":
+		key = fmt.Sprintf("%s/host/%s/peer_%s", CALICO_BGP, os.Getenv(HOSTNAME), version)
+	default:
+		return nil, fmt.Errorf("invalid neighbor type: %s", neighborType)
+	}
+	res, err := api.Get(context.Background(), key, &etcd.GetOptions{Recursive: true})
+	if errorButKeyNotFound(err) != nil {
+		return nil, err
+	}
+	if res == nil {
+		return nil, nil
+	}
+	ns := make([]*bgpconfig.Neighbor, 0, len(res.Node.Nodes))
+	for _, node := range res.Node.Nodes {
+		if n, err := getNeighborConfigFromPeer(node, neighborType); err != nil {
+			return nil, err
+		} else {
```

Comment:
you can probably get rid of this `else` since it returns when the if condition is true. 
And you can move the variable declaration to its own line for the scope to be reachable to `ns= append(ns,n)`

I'm not too worried about that, though.

Ref: https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow